File: euler145.scala
#! /usr/bin/env scala
!#

import math.{log10,floor,pow}

class MyInt(private var n:Int) extends java.lang.Number with Iterable[Int]{

	def length() = floor(log10(n)).toInt

	override def iterator(): Iterator[Int] = digits().iterator

	def digits(n:Int = n):Stream[Int] = {
		n match {
			case 0 => Stream.empty
			case _ => Stream.cons(n%10, digits(n/10))
		}
	}
	
	def reverse(n:Int = n, numDigits:Int = length):Stream[Int] = {
		numDigits >= 0 match{
			case true => Stream.cons((n / pow(10, numDigits).toInt) % 10, reverse(n, numDigits-1))
			case false => Stream.empty
		}
	}
	
	def intValue = n
	def doubleValue = n
	def floatValue = n
	def longValue = n
}

def reversible(d: Stream[Int], dr: Stream[Int], rem: Int = 0): Boolean = {
	(d, dr) match{
		case (Stream(), Stream()) => true
		case (dh#::dt, drh#::drt) => {
			var newd = dh+drh+rem
			newd%2!=0 && reversible(dt, drt, newd/10)
		}
	}
}

def solve(n: Int = 1000): Int = {
	var count = 0
	var j: MyInt = new MyInt(0)
	for (i <- 1 to n if i%10!=0){
		j = new MyInt(i)
		if (reversible(j.digits(), j.reverse())){
			count+=1
		}
	}
	return count
}


var a:Long=0
var b:Long=0
a=System.currentTimeMillis()
solve(pow(10,7).toInt)
b=System.currentTimeMillis()
print((b-a)/1000.0)

File: euler145.hs
#!/usr/bin/env runhaskell
import Control.Parallel (par)

reversible :: Int -> Bool
reversible a = reversible' da dar 0
	where
	da = digits a
	dar = reverse da -- (reverse da) needs to get first the "innermost" element -> becames unnecessarily strict
	
reversible' :: [Int] -> [Int] -> Int -> Bool
reversible' [] [] _ = True 
reversible' (d:da) (dr:dar) rem = (odd newd) && (reversible' da dar (newd `div` 10))
	where
	newd = d+dr+rem 

digits :: Int -> [Int]
-- digits = map (\c -> read [c]) . show
digits a
	| a' /= 0 = a `mod` 10 : digits a'
	| otherwise = [a]
	where
		a' = a `div` 10
		
reverse_digits :: Int -> [Int]
reverse_digits a
	| numdigits > 0 = d:(reverse_digits' a (numdigits-1))
	| otherwise = [a]
	where 
	numdigits = floor (logBase 10 (fromIntegral a))
	d = a `div` (10 ^ numdigits)
	
reverse_digits' :: Int -> Int -> [Int]
reverse_digits' a numdigits
	| numdigits > 0 = (d `mod` 10):(reverse_digits' a (numdigits-1))
	| otherwise = [a `mod` 10]
	where
	d = a `div` (10 ^ numdigits)
	
		
solve :: Int -> Int
solve 1 = 0
solve n 
	| n `mod` 10 /= 0 && reversible n = (solve (n-1) +) $! 1
	| otherwise = solve (n-1)
	
maxval = 10000000

solve' :: Int -> Int
solve' 1 = (solve 2) `par` ((solve 3) `par` ((solve 2)+(solve 3)) )
solve' n
	| n >= maxval = 0
	| n `mod` 10 /= 0 && reversible n = (1 +) $! solve (n+2)
	| otherwise = solve (n+2)
	
main :: IO()
main = print (solve' 1)
